[SPARK-40777][SQL][PROTOBUF] Protobuf import support and move error-classes.

SandishKumarHN opened the pull request:
This is the follow-up PR to https://github.com/apache/spark/pull/37972 and https://github.com/apache/spark/pull/38212

### What changes were proposed in this pull request?
1. Move spark-protobuf error classes to the spark error-classes framework(core/src/main/resources/error/error-classes.json). 
2. Support protobuf imports
3. validate protobuf timestamp and duration types.  


### Why are the changes needed?
N/A

### Does this PR introduce _any_ user-facing change?
None

### How was this patch tested?
Existing tests should cover the validation of this PR.  

CC: @rangadi @mposdev21 @gengliangwang

SandishKumarHN committed:
review changes nit

SandishKumarHN committed:
adding spark-protobuf error class into spark error-class framework

SandishKumarHN committed:
adding spark-protobuf error class into spark error-class framework

SandishKumarHN committed:
adding spark-protobuf error class into spark error-class framework, import support

SandishKumarHN committed:
Merge remote-tracking branch 'remote-spark/master' into SPARK-40777

# Conflicts:
#	connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala
#	connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala

SandishKumarHN committed:
adding spark-protobuf error class into spark error-class framework, import support

SandishKumarHN committed:
Merge branch 'SPARK-40777' into SPARK-40777-ProtoErrorCls

SandishKumarHN committed:
spark-protobuf error clss frameworks, import support, timestamp and duration validation

SandishKumarHN commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -62,14 +63,16 @@ object SchemaConverters {
       case BYTE_STRING => Some(BinaryType)
       case ENUM => Some(StringType)
-      case MESSAGE if fd.getMessageType.getName == "Duration" =>
+      case MESSAGE if fd.getMessageType.getName == "Duration" &&
```
@rangadi My assumption is that users should be able to use the Timestamp and Duration message types with different fields.

SandishKumarHN commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -196,27 +195,32 @@ private[sql] object ProtobufUtils extends Logging {
     }
 
-    val descriptorProto: DescriptorProtos.FileDescriptorProto = fileDescriptorSet.getFile(0)
+    val descriptorProto: DescriptorProtos.FileDescriptorProto =
```
@rangadi This handles the protobuf import; please let me know if you know of a better approach.

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
cc @MaxGekk FYI

SandishKumarHN committed:
fixing typos

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3210,4 +3211,183 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+    protobufType: String,
+    sqlType: String): Throwable = {
+    new AnalysisException(
+      errorClass = "_LEGACY_ERROR_TEMP_2251",
```
We use the prefix `_LEGACY_ERROR_TEMP_` to convert existing exception to error classes but for new exception, please, assign appropriate names to error classes.

MaxGekk commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala:
```diff
@@ -244,7 +239,13 @@ private[sql] class ProtobufDeserializer(
+          toFieldStr(catalystPath),
+          s"${protoType} ${protoType.toProto.getLabel} ${protoType.getJavaType}" +
+            s" ${protoType.getType}",
+          catalystType.sql)
```
Pass `catalystType` into `cannotConvertProtobufTypeToSqlTypeError()`, and use `toSQLType` to quote the type.

SandishKumarHN replied:
done

rangadi commented on connector/protobuf/pom.xml:
```diff
@@ -123,6 +123,7 @@
             <configuration>
               <protocArtifact>com.google.protobuf:protoc:${protobuf.version}</protocArtifact>
               <protocVersion>${protobuf.version}</protocVersion>
+              <includeMavenTypes>direct</includeMavenTypes>
```
Could you add comment here?
Is this for protos like `Timestamp' ? Looks like our code handles any Timestamp defined similar com.google.protobuf.Timestamp. Ok to not to include these.

SandishKumarHN replied:
@rangadi this protoc maven option is equal to CLI "protoc --include_imports" and I have added validations to protobuf Timestamp and Duration types. and currently, maven protoc does not generate classes for google protobuf types. sbt is enabled by default.

rangadi replied:
I see. What would fail if you remove this?

SandishKumarHN replied:
@rangadi just "messageClassName" param feature unit tests(Timestamp and Duration). The protoc plugin is only used for the test right?

rangadi replied:
That's because this removes existing 'Timestamp' message : https://github.com/apache/spark/pull/38344/files#diff-97aac63266f3c60eef9bd8dd1b76be3a5bd77fe4d17fa6fa370f5e0d9428a0a9L172

We could undo that change.

SandishKumarHN replied:
@rangadi how about supporting imports? I made this [change](https://github.com/apache/spark/pull/38344/files#diff-97aac63266f3c60eef9bd8dd1b76be3a5bd77fe4d17fa6fa370f5e0d9428a0a9L172) of "Timestamp" and "Duration" to show a unit test for imports.

SandishKumarHN replied:
@rangadi how about splitting Timestamp and Duration into our own .proto files and then importing them in the functions_suite.proto file?

rangadi replied:
Yep. That would be nice. That will mimic normal proto files organization.

SandishKumarHN replied:
@rangadi made changes same.

SandishKumarHN committed:
review changes, name protobuf error class and comment in pom.xml

SandishKumarHN committed:
review chages rm includeMavenTypes and import own protos

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -196,27 +194,31 @@ private[sql] object ProtobufUtils extends Logging {
     }
 
-    val descriptorProto: DescriptorProtos.FileDescriptorProto = fileDescriptorSet.getFile(0)
+    val descriptorProto: DescriptorProtos.FileDescriptorProto =
```
could you some brief comments here?
Is the last file the right file?

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -196,27 +194,31 @@ private[sql] object ProtobufUtils extends Logging {
+    val descriptorProto: DescriptorProtos.FileDescriptorProto =
+      fileDescriptorSet.getFileList.asScala.last
+
+    var fileDescriptorList = List[Descriptors.FileDescriptor]()
```
Is this the import file list? What happens when the imported file imports other files? i.e. A imports B and B imports C.

SandishKumarHN replied:
@rangadi some changes were made to support nested imports. Reading protobuf descriptors from the bottom up, the last element in FileDescriptorSet is the initial FileDescriptorProto, from which we will continue to find more FileDescriptors recursively.

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:
```diff
@@ -62,14 +63,18 @@ object SchemaConverters {
-        //        google.protobuf package should default to corresponding Catalylist types.
+      case MESSAGE
+        if fd.getMessageType.getName == "Timestamp" &&
+        fd.getMessageType.getFields.size() == 2 &&
```
Could move all the conditions into braces (a & b & ...)?

SandishKumarHN replied:
fixed

SandishKumarHN committed:
review changes: nested import support, nit, build fix

SandishKumarHN commented on connector/protobuf/src/test/resources/protobuf/timestamp.proto:
```diff
@@ -0,0 +1,26 @@
+message Timestamp {
+  int64 seconds = 1;
+  int32 nanos = 2;
+}
```
done

SandishKumarHN commented on connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala:
```diff
@@ -163,18 +163,22 @@ class ProtobufSerdeSuite extends SharedSparkSession {
     assert(e.getMessage === expectMsg)
-    assert(e.getCause.getMessage === expectedCauseMessage)
+    if (e.getCause != null) {
+      assert(e.getCause.getMessage === expectedCauseMessage)
```
made changes to support new error class frameworks.

rangadi reviewed (approved):
Thanks for the updates. Suggested some scala style changes. 
LGTM for protobuf functional changes. 
We can get @MaxGekk to approve error message related changes.

SandishKumarHN committed:
review changes style changes, nit

SandishKumarHN commented:
@HyukjinKwon getting below error in build any idea? nothing related PR

```
File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line [68](https://github.com/SandishKumarHN/spark/actions/runs/3325103283/jobs/5498895243#step:17:69)0, in _load_unlocked
  File "/usr/local/lib/python3.9/dist-packages/_pytest/assertion/rewrite.py", line 168, in exec_module
    exec(co, module.__dict__)
  File "/usr/local/lib/python3.9/dist-packages/pytest_mypy_plugins/collect.py", line 13, in <module>
    from py._path.local import LocalPath
ModuleNotFoundError: No module named 'py._path'; 'py' is not a package
```

SandishKumarHN committed:
Merge remote-tracking branch 'remote-spark/master' into SPARK-40777-ProtoErrorCls

SandishKumarHN commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3210,4 +3211,189 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+    new AnalysisException(
+      errorClass = "SQL_TYPE_TO_PROTOBUF_TYPE_ERROR",
+      messageParameters = Map(
+        "sqlColumn" -> sqlColumn,
```
@MaxGekk made changes.

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -196,30 +194,52 @@ private[sql] object ProtobufUtils extends Logging {
+  private def createDescriptorProtoMap(
+    fileDescriptorSet: FileDescriptorSet): Map[String, FileDescriptorProto] = {
+    fileDescriptorSet.getFileList().asScala.map { descriptorProto =>
+      descriptorProto.getName() -> descriptorProto
```
Btw, What happens if there are two different files with the same name? E.g. 
`protoc ... a/b/events.proto x/y/events.proto`

SandishKumarHN replied:
@rangadi protoc command would throw the below error, protoc won't allow duplicate imports. 

```
protoc --java_out=./ functions_suite.proto
basicmessage_1.proto:30:9: "org.apache.spark.sql.protobuf.protos.BasicMessage.id" is already defined in file "basicmessage.proto".
basicmessage_1.proto:31:10: "org.apache.spark.sql.protobuf.protos.BasicMessage.string_value" is already defined in file "basicmessage.proto".
basicmessage_1.proto:32:9: "org.apache.spark.sql.protobuf.protos.BasicMessage.int32_value" is already defined in file "basicmessage.proto".
basicmessage_1.proto:33:9: "org.apache.spark.sql.protobuf.protos.BasicMessage.int64_value" is already defined in file "basicmessage.proto".
basicmessage_1.proto:34:10: "org.apache.spark.sql.protobuf.protos.BasicMessage.double_value" is already defined in file "basicmessage.proto".
```

rangadi replied:
The command is using only single proto file. Not the same as the example above, right?

SandishKumarHN replied:
@rangadi yeah correct but functions_suite.proto is including both basicmessage.proto and basicmessage_1.proto, wouldn't it equal the same as above? 
```
import "timestamp.proto";
import "duration.proto";
import "basicmessage.proto";
import "basicmessage_1.proto";
```

rangadi replied:
That is not what I am referring to. This is not really a problem since it won't even compile. 

Notice two files as input for `protoc` above. May be that will since `name` will include the path.

SandishKumarHN replied:
@rangadi Yes, protoc allowed two .proto files from different paths, but the last .proto file overrides the first one if there are any conflicts. but It is entirely up to the user to determine how to create a .desc file right?

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -196,30 +194,52 @@ private[sql] object ProtobufUtils extends Logging {
-        new Array[Descriptors.FileDescriptor](0))
+      val fileDescriptorProtoIndex = createDescriptorProtoMap(fileDescriptorSet)
+      val fileDescriptor: Descriptors.FileDescriptor =
+        buildFileDescriptor(fileDescriptorSet.getFileList.asScala.last, fileDescriptorProtoIndex)
```
There is one more related issue here: We can find messages defined in this file, but not the message defined inside the imported file. 

I think it is better to fix all of this properly, since you are already working on this. 
I.e. we should be able to use any message defined in the descriptor file. 
Essentially we can build `List [ FileDescriptor ]` and search for user's message in all of those. We could give preference for `last` if in this search order.

SandishKumarHN replied:
@rangadi I removed this option on purpose; since the user would pass the main descriptor file as a parameter, we should just look at the main descriptor. What if the user has the same MessageName in multiple .proto files? Which one should I read?

the descriptor structure is first imports, then message definition, with this assumption "last" would be the main message definition. 

how about looking at the main message definition first, if don't find we can look at the "imports" return as soon as the messagename matches.?

rangadi replied:
I think that is pretty restrictive. There will be cases where user wants to use the protobuf defined in imported files. E.g. they might have one proto file jus tot import bunch of other file that define the protobufs. 

What are we gaining by having restriction to allow messages only in the last listed file? In addition there could be be multiple top level files (i.e. `protoc` command is run with multiple files).

I think fixing it properly would be the right thing. We can level a TODO here, I can add that in a follow up. LMK.

SandishKumarHN replied:
@rangadi I just added a basic working method for searching messageName through all imports. and left a TODO to revisit. and also made changes to unit tests(ProtobufFunctionsSuite) to use a single .desc file

rangadi reviewed (changes requested):
Commented on one more case we need to improve while building File descriprors.

SandishKumarHN committed:
review changes: search messages on all imports

MaxGekk commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDataToCatalyst.scala:
```diff
@@ -98,13 +92,9 @@ private[protobuf] case class ProtobufDataToCatalyst(
+        throw QueryCompilationErrors.malformedRecordsDetectedInRecordParsingError(e)
       case _ =>
-        throw new AnalysisException(unacceptableModeMessage(parseMode.name))
+        throw QueryCompilationErrors.parseModeUnsupportedError("from_protobuf", parseMode)
```
ditto: prettyName

SandishKumarHN replied:
@MaxGekk fixed.

SandishKumarHN committed:
review changes: use prettyName

SandishKumarHN committed:
Merge remote-tracking branch 'remote-spark/master' into SPARK-40777-ProtoErrorCls

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3210,4 +3211,190 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+    new AnalysisException(
+      errorClass = "CATALYST_TYPE_TO_PROTOBUF_ENUM_TYPE_ERROR",
+      messageParameters = Map(
+        "sqlColumn" -> sqlColumn,
```
If it is a column id, wrap it by `toSQLId()`

MaxGekk commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufSerializer.scala:
```diff
@@ -215,10 +217,12 @@ private[sql] class ProtobufSerializer(
-            s"schema is incompatible (sqlType = ${catalystType.sql}, " +
-            s"protoType = ${fieldDescriptor.getJavaType})")
+        throw QueryCompilationErrors.cannotConvertCatalystTypeToProtobufTypeError(
+          toFieldStr(catalystPath),
```
Please, don't quote it twice if you do that inside of `cannotConvertCatalystTypeToProtobufTypeError()` by `toSQLId`

SandishKumarHN replied:
okay, will call toSQLType inside the error class, avoid using toFieldStr for sqlType

SandishKumarHN committed:
review changes: nit, option, runtime error

SandishKumarHN commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3210,4 +3211,190 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+      messageParameters = Map(
+        "protobufType" -> protobufType,
+        "toType" -> toSQLType(sqlType)),
+      cause = Some(cause.getCause))
```
fixed

SandishKumarHN commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3210,4 +3211,190 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+    new AnalysisException(errorClass = "INVALID_BYTE_STRING_ERROR", messageParameters = Map.empty)
+  }
+
+  def malformedRecordsDetectedInRecordParsingError(cause: Throwable): Throwable = {
```
it is runtime, moved to ExecutionError class.

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -792,6 +887,21 @@
       "Unable to acquire <requestedBytes> bytes of memory, got <receivedBytes>"
     ]
   },
+  "UNABLE_TO_CONVERT_TO_PROTOBUF_TYPE" : {
```
Same as above?

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -792,6 +887,21 @@
+      "Unable to convert SQL type <toType> to Protobuf type <protobufType>."
+    ]
+  },
+  "UNABLE_TO_LOCATE_PROTOBUF_MESSAGE_ERROR" : {
```
```suggestion
  "PROTOBUF_MESSAGE_NOT_FOUND" : {
```

SandishKumarHN replied:
@srielau fixed all the error class names

SandishKumarHN commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -706,6 +736,66 @@
+      "Unable to convert <protobufType> of Protobuf to SQL type <toType>."
+    ]
+  },
+  "PROTOBUF_TYPE_TO_SQL_TYPE_ERROR" : {
```
this is thrown when failing to convert protobuf fields into SQL type.

SandishKumarHN commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -706,6 +736,66 @@
+      "Protobuf type not yet supported: <protobufType>."
+    ]
+  },
+  "PROTOBUF_TYPE_TO_CATALYST_TYPE_ERROR" : {
```
this is thrown when failing to convert the protobuf message to SQL struct type.

SandishKumarHN committed:
error class name changes, more details to error message

error class name changes, more details to error message

SandishKumarHN commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -65,6 +70,11 @@
   },
+  "CATALYST_TYPE_TO_PROTOBUF_ENUM_TYPE_ERROR" : {
+    "message" : [
+      "Cannot convert SQL <sqlColumn> to Protobuf <protobufColumn> because <data> cannot be written since it's not defined in ENUM <enumString>"
```
@srielau there is no ENUM type in spark, so we convert the protobuf enum to spark string.  so we are trying to throw an enum-specific error here.

SandishKumarHN commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -706,6 +736,66 @@
+      "Could not find dependency: <dependencyName>"
+    ]
+  },
+  "PROTOBUF_DESCRIPTOR_ERROR" : {
```
@srielau Along with the message added the pathname of the descriptor file and the same for the one below. Aside from the error message, we don't have much information at this time.

SandishKumarHN committed:
Merge remote-tracking branch 'remote-spark/master' into SPARK-40777-ProtoErrorCls

SandishKumarHN committed:
NO_UDF_INTERFACE_ERROR to NO_UDF_INTERFACE

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -177,47 +175,75 @@ private[sql] object ProtobufUtils extends Logging {
   }
 
-  private def parseFileDescriptor(descFilePath: String): Descriptors.FileDescriptor = {
+  private def parseFileDescriptor(descFilePath: String): List[Descriptors.FileDescriptor] = {
```
Rename to `parseFileDescriptorSet` (otherwise it sounds like it is parsing just one file descriptor).

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -177,47 +175,75 @@ private[sql] object ProtobufUtils extends Logging {
   def buildDescriptor(descFilePath: String, messageName: String): Descriptor = {
-    val descriptor = parseFileDescriptor(descFilePath).getMessageTypes.asScala.find { desc =>
-      desc.getName == messageName || desc.getFullName == messageName
+    val descriptorList = parseFileDescriptor(descFilePath).map(fileDescriptor => {
```
Style: use `find()` rather than map().filter(). 

(you can use `findLast()` if there is a reason to use the last match).

SandishKumarHN replied:
@rangadi makes sense to use find and return, fixed.

SandishKumarHN committed:
review changes scala style, find, parseFileDescriptorSet

SandishKumarHN committed:
Merge remote-tracking branch 'remote-spark/master' into SPARK-40777-ProtoErrorCls

rangadi commented on connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/ProtobufUtils.scala:
```diff
@@ -178,46 +176,73 @@ private[sql] object ProtobufUtils extends Logging {
-      desc.getName == messageName || desc.getFullName == messageName
-    }
+    val fileDescriptor = parseFileDescriptorSet(descFilePath)
+      .find(!_.getMessageTypes.asScala.find(desc =>
```
Oh, we need to find() invocations and check is made twice. How about this? [Optional to do, only a suggestion]

     // Find the first message descriptor that matches the name.
     val descriptorOpt =  parseFileDescriptorSet(descFilePath)
         .flatMap { fileDesc => 
                fileDesc.getMessageTypes.asScala.find { desc => 
                      desc.getName == messageName || desc.getFullName == messageName
                }
         }.headOption

      descriptoOpt match {
         case Some(d) => d
         case None => throw QueryCompilationErrors.unableToLocateProtobufMessageError(messageName)
     }

SandishKumarHN replied:
fixed

rangadi reviewed (approved):
LGTM on import functionality changes. Will let other reviewers to approve the error message changes.

SandishKumarHN committed:
review changes buildDescriptor suggested changes

srielau commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3212,4 +3212,181 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+    val message = if (hasDots) "" else ". Ensure the class name includes package prefix."
+    new AnalysisException(
+      errorClass = "CANNOT_LOAD_PROTOBUF_CLASS",
+      messageParameters = Map("protobufClassName" -> protobufClassName, "message" -> message),
```
We don't want text as a parameter. Aside to make this text depends on whether there is any dot seems to be quite subtle. Maybe it simply doesn't have enough dots?
I propose to move this text into the error message (unconditionally).

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -742,6 +832,11 @@
     ],
     "sqlState" : "22023"
   },
+  "SQL_TYPE_TO_PROTOBUF_ENUM_TYPE_ERROR" : {
```
Rename without _ERROR

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -688,6 +733,51 @@
+      "Type mismatch encountered for field: <field>"
+    ]
+  },
+  "PROTOBUF_FIELD_TYPE_TO_SQL_TYPE_ERROR" : {
```
How about:
```suggestion
  "CANNOT_CONVERT_PROTOBUF_FIELD_TYPE_TO_SQL_TYPE" : {
```

srielau commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -29,12 +44,22 @@
   },
+  "CANNOT_LOAD_PROTOBUF_CLASS" : {
+    "message" : [
+      "Could not load Protobuf class with name <protobufClassName><message>"
```
```suggestion
      "Could not load Protobuf class with name <protobufClassName>. Ensure the class name includes package prefix."
```

SandishKumarHN replied:
@srielau fixed the suggested changes.

SandishKumarHN committed:
review changes, error classes

srielau reviewed (approved):
LGTM

rangadi commented:
@MaxGekk are you still reviewing this? 
@SandishKumarHN is there any more review to be addressed? If we are ready, I can ask @HeartSaVioR to merge this (before his weekend starts in Seoul :) ).

SandishKumarHN commented:
> @MaxGekk are you still reviewing this? @SandishKumarHN is there any more review to be addressed? If we are ready, I can ask @HeartSaVioR to merge this (before his weekend starts in Seoul :) ).

@rangadi I have addressed all of the review comments.

MaxGekk reviewed (commented):
Error classes LGTM

HeartSaVioR reviewed (approved):
+1

HeartSaVioR commented:
Thanks! Merging to master.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3212,4 +3212,179 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+  def descrioptorParseError(descFilePath: String, cause: Throwable): Throwable = {
+    new AnalysisException(
+      errorClass = "CANNOT_PARSE_PROTOBUF_DESCRIPTOR",
+      messageParameters = Map.empty("descFilePath" -> descFilePath),
```
Accidentally found this. Is `messageParameters` set to an empty map, if so, this will fail in runtime w/ an internal error. Could you write a test which trigger the error and fix the bug.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -3212,4 +3212,179 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
+  def failedParsingDescriptorError(descFilePath: String, cause: Throwable): Throwable = {
+    new AnalysisException(
+      errorClass = "CANNOT_CONSTRUCT_PROTOBUF_DESCRIPTOR",
+      messageParameters = Map.empty("descFilePath" -> descFilePath),
```
@SandishKumarHN Please, fix this. Map.empty -> Map
